core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingInterceptor.java

```diff
@@ -140,119 +131,90 @@ public final Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEnt
+   private void handleGetAllCommandReturn(InvocationContext rCtx, VisitableCommand rCommand, Object rv, Throwable t) throws Throwable {
+      GetAllCommand getAllCommand = (GetAllCommand) rCommand;
+      if (t != null) {
+         if (t instanceof RpcException && performPartitionCheck(rCtx, getAllCommand)) {
+            // We must have received an AvailabilityException from one of the owners.
+            // There is no way to verify the cause here, but there isn't any other way to get an invalid
+            // get response.
+            throw log.degradedModeKeysUnavailable(((GetAllCommand) rCommand).getKeys());
+         } else {
+            throw t;
          }
+      }
 
-         if (performPartitionCheck(rCtx, getAllCommand)) {
-            // We do the availability check after the read, because the cache may have entered degraded mode
-            // while we were reading from a remote node.
-            for (Object key : getAllCommand.getKeys()) {
-               partitionHandlingManager.checkRead(key);
-            }
-
-            Map<Object, Object> result = ((Map<Object, Object>) rv);
-            // If all owners left and we still haven't received the availability update yet, we could return
-            // an incorrect value. So we need a special check for missing results.
-            if (result.size() != getAllCommand.getKeys().size()) {
-               // Unlike in PartitionHandlingManager.checkRead(), here we ignore the availability status
-               // and we only fail the operation if _all_ owners have left the cluster.
-               // TODO Move this to the availability strategy when implementing ISPN-4624
-               Set<Object> missingKeys = new HashSet<>(getAllCommand.getKeys());
-               missingKeys.removeAll(result.keySet());
-               for (Iterator<Object> it = missingKeys.iterator(); it.hasNext(); ) {
-                  Object key = it.next();
-                  List<Address> readOwners = distributionManager.getCacheTopology().getDistribution(key).readOwners();
-                  if (InfinispanCollections.containsAny(transport.getMembers(), readOwners)) {
-                     it.remove();
-                  }
-               }
-               if (!missingKeys.isEmpty()) {
-                  throw log.degradedModeKeysUnavailable(missingKeys);
-               }
-            }
+      if (performPartitionCheck(rCtx, getAllCommand)) {
+         // We do the availability check after the read, because the cache may have entered degraded mode
+         // while we were reading from a remote node.
+         for (Object key : getAllCommand.getKeys()) {
+            partitionHandlingManager.checkRead(key);
          }
 
-         // TODO We can still return a stale value if the other partition stayed active without us and we haven't entered degraded mode yet.
-      });
+         Map<Object, Object> result = ((Map<Object, Object>) rv);
+         if (result.size() != getAllCommand.getKeys().size()) {
```

Comment:
That comment addresses the situation when the entry is available but the value is not in the cache; in that case the map contains the key -> null mapping. The mapping is not in the map when some of the keys are actually non-available as all owners have been lost.